crypto/x509/x_all.c

```diff
@@ -70,8 +70,11 @@ int X509_verify_ex(X509 *a, EVP_PKEY *r, OPENSSL_CTX *libctx, const char *propq)
     EVP_MD_CTX *ctx = NULL;
     ASN1_OCTET_STRING *id = NULL;
 
-    if (X509_ALGOR_cmp(&a->sig_alg, &a->cert_info.signature))
-        return 0;
+    /* Check signature alg consistency according to RFC 5280 section 4.1.1.2 */
+    if (X509_ALGOR_cmp(&a->sig_alg, &a->cert_info.signature) != 0) {
+        X509err(0, X509_R_SIGNATURE_ALGORITHM_MISMATCH);
+        return -1;
```

Comment:
Is the motivation for the change in return code here that the caller should not have called the function in the first place, given that the algorithms don't match?

That sounds plausible, BUT we promise to return zero when thing `a` is not signed by key `b`, and up till now callers have been allowed to call us even with an algorithm mismatch, and we get to say, that `b` did not sign `a`.  The negative results are for situations where we could even determine the answer, and that's simply not the case here.

So I rather think the return value must remain zero in this case, lest we break the established API semantics.

Reply:
> Is the motivation for the change in return code here that the caller should not have called the function in the first place, given that the algorithms don't match?

Let's say: the motivation is similar to that.
As I wrote in the comment and in the respective commit message, a cert with inconsistent sig alg fields should must not have been produced at all. 
This change simply adds a check for a cert being ill-formed in that regard and throws an 'internal' error (-1) in this case.
Note that in such a case it does not state whether the cert signature is valid (1) or not (0).

> That sounds plausible, BUT we promise to return zero when thing `a` is not signed by key `b`, and up till now callers have been allowed to call us even with an algorithm mismatch, and we get to say, that `b` did not sign `a`. The negative results are for situations where we could even determine the answer, and that's simply not the case here.
> 
> So I rather think the return value must remain zero in this case, lest we break the established API semantics.

Not really. See the documentation of `X509_verify_ex()` and friends, which now lives in `doc/man3/X509_verify.pod`:
```
X509_verify_ex(), X509_REQ_verify_ex(), X509_REQ_verify() and X509_CRL_verify()
return 1 if the signature is valid and 0 if the signature check fails.
If the signature could not be checked at all because it was ill-formed
or some other error occurred then -1 is returned.
```
and this PR does not indend - and in fact does not - change the semantics of their return code.
Also note that `ASN1_item_verify_ctx()` called next in `X509_verify_ex`, returns `-1` on internal errors,
so callers already need to be prepared for that.

Reply:
I am going to hold firm (unless there's clear support to the contrary from other OTC members) on the `0` in this case.  To an API caller, the purpose of `0` vs.`-1` is to express the difference between a definitive answer (definitely does not match) and an internal error, where OpenSSL could not decide whether there or isn't a match (out of memory, malformed structure to the point that we could not parse it to find a signature, ...).  Here, we're sure that the signature does not match (wrong algorithm or payload).  So the return value is 0.

Reply:
@vdukhovni, due to your last comment I had a more holistic look at the `crypto/x509/{x_all,x509_vfy,x3_purp}.c` code again and found that my addition of the check `X509_ALGOR_cmp(&a->sig_alg, &a->cert_info.signature) != 0` should anyway not be in `X509_verify_ex()`. So I removed it from there and we do not need to discuss further which would be the right return code in this case.

It turns out that the signature alg consistency check makes most sense as part of `X509v3_cache_extensions()`, which besides its (slightly misleading or at least insufficient) name already does quite a number of checks regarding cert data validity.
So I moved the new test there.

Doing so I also found that `X509v3_cache_extensions()` contained various issues, which I fixed:
* coding style nits
* in case `EVP_MD_fetch(libctx, "SHA1", propq)` returns NULL 
this must not be recorded as `x->ex_flags |= EXFLAG_INVALID`
* no error queue entry was generated in case the certs is invalid
* commenting was a little inconsistent/incomplete
* the call `x509_init_sig_info(x)` might fail internally, but due to its `void` return type the calling function cannot handle this properly. Since correcting this would require more changes also to `crypto/x509/x509_set.c` I just added TODOs indicating where improvements should be done
* documentation did not mention that also information not related to X.509v3 extensions is cached

I've rebased the PR and make the commit containing the changes just described last.
So when having a look at the PR again you can focus on that: https://github.com/siemens/openssl/commit/763d4339f633355265fd8447c21220718b6ff3ab.

In case further discussion is needed on this commit I would move it to a separate PR such that the changes more directly related to fixing #1418 are not further held up from getting approved and merged.

Reply:
Is this okay now?

BTW, Travis invocation currently tends to get stuck on many PRs, 
but the tests work fine for me locally. I've just rebased again to give it a kick.

Reply:
I see where you're going with this, but I'm even less convinced that the signature algorithm check belongs in `x509_cache_extensions()`. That function's job is to interpret, not verify, the certificate. It should only fail when the certificate is malformed to the point of not being parseable.  If we can parse it, we cache the results and return success!  It is up to code that's validating the certificate signature to decide whether the signature is *valid*, which is distinct from valid, which is turn distinct from "some internal error prevented a definite outcome".

So please restore the 0 zero return value, and allow certificates with a signature algorithm mismatch to be parsed. With DANE-EE(3) they're largely just public-key containers, and their signature is irrelevant, but still honour the key usage (though it is no longer cryptographically protected, mostly perhaps it is pain not to do that...).  The only think that should fail for such certificates is signature validation, and it should fail with a zero return.

Reply:
Thanks for your further comments.

The change to `X509_verify_ex()` still present in `x_all.c` was by mistake; I've revered that restoring the original behavior.

I've also removed the signature consistency check that I moved/added to `X509v3_cache_extensions()` yesterday.
So in that function I only left the improved/extended comments, the fix of the case `sha1 == NULL` (which so far lead to the wrong assertion that the cert is invalid), and the coding style improvements.

See the new commit with the partly reverted changes: https://github.com/siemens/openssl/commit/786269509f00dfe8f3f961e14b9296f9c00e0fda